sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/ddl.scala

```diff
@@ -90,17 +90,31 @@ case class RefreshResource(path: String)
 /**
  * Builds a map in which keys are case insensitive
  */
-class CaseInsensitiveMap(map: Map[String, String]) extends Map[String, String]
+class CaseInsensitiveMap[T] private(baseMap: Map[String, T]) extends Map[String, T]
```

Comment:
If we decide to go case-insensitive all usage of options, probably we don't need class CaseInsensitiveMap adapter class any longer.  Instead, we should probably convert to user provided options to lower case directly.